README.md
## DALY

[![Travis-CI Build Status](https://travis-ci.org/brechtdv/DALY.svg?branch=master)](https://travis-ci.org/brechtdv/DALY)
[![CRAN_Status_Badge](http://www.r-pkg.org/badges/version/DALY)](http://cran.r-project.org/package=DALY)
[![Cranlogs Downloads](http://cranlogs.r-pkg.org/badges/grand-total/DALY)](http://cran.r-project.org/package=DALY)

The DALY Calculator is a free, open-source Graphical User Interface for probabilistic disability-adjusted life year (DALY) calculation, developed in the R environment for statistical computing.

The main goal of the DALY Calculator is to provide a flexible and easy-to-use tool for DALY calculation and to promote consistency in DALY uncertainty analysis.

#### Install

To download and install the latest released version from [CRAN](https://cran.r-project.org/package=DALY):
```r
install.packages("DALY")
```

To download and install the latest development version from GitHub:
```r
devtools::install_github("brechtdv/DALY")
```

#### More information
More information and tutorials are available at http://daly.cbra.be/.


NEWS
DALY 1.5.0
==========

PERFORMANCE IMPROVEMENTS

* the GBD2010 and WHO/GHE standard life expectancy tables have been added

* the GBD2010 standard life expectancy table is now the default option

BUG FIXES

* fixed imports declarations in NAMESPACE

* converted `round()` to `fround()` in `DALY.cpp`

* `sensitivity` functions could result in errors due to loss of dimensions (@thanhleviet)


DALY 1.4.0
==========

BUG FIXES

* fixed NAMESPACE issues with imported packages

* removed Poisson distribution on number of cases and deaths

* fixed bug with scientific notation in Tk tables


DALY 1.3.0
==========

NEW FEATURES

* introduction of a `DALY_list()` function to combine different `DALY` objects

* introduction of a `plot()` method for objects of class `DALY` and `DALY_list`

* introduction of a `scatterplot()` method for objects of class `DALY` and
  `DALY_list`

* introduction of a `print()` and `plot()` method for objects of class
  `DALY_sensitivity`

PERFORMANCE IMPROVEMENTS

* function `sensitivity()` has been updated to allow calculating
  (standardized) regression coefficients and partial correlation
  coefficients, either on the original or the ranked input values

BUG FIXES

* fixed mismatch between Authors@R and Author fields in DESCRIPTION file

* the use of extern "C" {} in the `DALY.cpp` file has been revised to fix
  issues with clang


DALY 1.2.0
==========

NEW FEATURES

* introduction of a `sensitivity()` function

* function `getDALY()` gains arguments `aw` and `dr`

PERFORMANCE IMPROVEMENTS

* the DALY Calculator output now also includes the simulated values from the 
  input distributions

* several functions have been updated to make proper use of the `match.arg()` 
  functionality

* the `DALY` database on the search path is replaced by a `.DALYenv` 
  environment in the DALY package

BUG FIXES

* the `readDALYdata()` function now correctly reads `.RData` files when 
  specified through the file argument

* all instances of `INC` and `MRT` have now consistently been renamed to 
  `cases` and `deaths`, respectively

* the DALY Calculator manual has been rebuilt to fix issues with pdffonts


DALY 1.1.0
==========

NEW FEATURES

* introduction of an `aggregate()` method for objects of class `DALY`

PERFORMANCE IMPROVEMENTS

* each window has a "save or cancel" policy

* the basic DALY Calculator output consists of simulated DALYs, YLLs, YLDs, 
  incident cases and deaths per outcome and per age/sex class

* `print()`, `summary()` and `hist()` methods have been updated

* saving and loading data is implemented through `.RData` files instead of 
  `.txt ` files

DEPRECATED AND DEFUNCT

* text files with input data can no longer be read by the DALY Calculator

* as a result, the DALY Calculator Excel template is defunct

INSTALLATION

* DALY 1.1.0 can be installed and run under Mac OS X, Linux and Windows 
  operating systems. The DALY Calculator is only launched after attaching the 
  package in an interactive session, which means that the package can be 
  installed without having the required tools installed. Loading the package 
  under Mac OS X and Linux requires the prior installation of the X Window 
  System (X11) and the Tcl/Tk toolkit.

BUG FIXES

* global variables removed; all variables needed by the DALY Calculator are 
  stored in a separate `DALY` database

* welcome messages implemented through `packageStartupMessage()`

* problems with installation of `BWidget` resolved by replacing 
  `tkwidget(parent, type = "ComboBox", ...)` by `ttkcombobox()`

* `set.seed()` can be used to select the seed used for the random number 
  generation processes within the DALY Calculator

* a same window can be opened only once; calling a window opening function 
  while the corresponding window is already opened will return focus to the 
  existing window

* evaluation of input data improved by using regular expressions

* empty cells in the *Life Expectancy* window produce an error message, while 
  they are still allowed in other windows

* `iterations` can only be set to a strictly positive integer value

* on closing the DALY Calculator main window, all other DALY Calculator 
  windows will close as well

* checkboxes may also be set by clicking the corresponding text.


DALY 1.0.0
==========

NEW FEATURES

* calculate DALYs, YLDs, YLLs and Deaths for a simple incidence-based disease 
  model

* able to include up to eight outcomes

* five age-groups included, i.e., 0-4; 5-14; 15-44; 45-59; 60+

* life expectancy table can be defined by user; standard table is the Coale 
  and Demeny model life-table West, level 26 and 25

* two examples included: Neurocysticercosis in Cameroon and 
  Toxoplasmosis in the Netherlands

* saving/loading data through .txt file

* the basic DALY Calculator output consists of simulated DALYs, YLLs, YLDs 
  and deaths per outcome, aggregated over age/sex classes

* `print()`, `summary()` and `hist()` methods for objects returned by the
  DALY Calculator.

DALY 0.0.0
==========

* A first version of the DALY Calculator was developed in November 2010, but 
  was never released. This version consisted of a single `data` window, and 
  was only able to reproduce the Neurocysticercosis example. 

hist.DALY.R
## 'hist' S3 Method for class 'DALY'
## based on 'grid' package

hist.DALY <-
function(x, xval = c("DALY", "YLD", "YLL", "cases", "deaths"),
         prob = 0.95, central = c("mean", "median"),
         breaks = 25, fill = "grey95", ...){

  ## Check 'xval'
  xval <- match.arg(xval)

  ## Check 'prob'
  if (is.null(prob) || is.na(prob) || !is.numeric(prob))
    stop("'prob' must be a numeric value between 0 and 1")
  if (prob < 0 | prob > 1)
    stop("'prob' must be a numeric value between 0 and 1")

  ## Check 'central'
  central <- match.arg(central)

  ## Obtain aggregated 'xval'
  y <- aggregate(x, by = "total")[xval][[1]]

  ## Plot settings
  h <- hist(y, breaks = breaks, plot = FALSE)
  h2 <- hist(h$counts, plot = FALSE)

  x_val <- (h$breaks - min(h$breaks)) / (max(h$breaks) - min(h$breaks))
  y_val <- (h$counts / max(h$counts)) * .95
  y_val <- c(t(cbind(y_val, y_val, 0)))

  x_lb <- seq(min(h$breaks), max(h$breaks), length.out = 5)
  x_at <- (x_lb - min(h$breaks)) / (max(h$breaks) - min(h$breaks))

  y_lb <- round(seq(0, max(h2$breaks), length.out = 5), 0)
  y_at <- y_lb / max(h$counts)
  if (max(y_at) > 1.05){
    y_lb <- y_lb[-5]
    y_at <- y_at[-5]
  }

  ## Obtain summary statistics of 'xval'
  s <- summarize(y, .prob = prob)
  q <- approx(h$breaks, seq(0, 1, length.out = length(h$breaks)), s)$y
  at <- c(q[3] / 2, mean(q[3:4]), q[4] + (1 - q[4]) / 2)

  ## Plot histogram
  grid.newpage()

  ## top strip
  pushViewport(viewport(y = unit(1, "npc") - unit(3.5, "lines"),
                        x = unit(0, "npc") + unit(4, "lines"),
                        width = unit(.95, "npc") - unit(4, "lines"),
                        height = unit(1, "lines"),
                        just = c("left", "top"), name = "strip"))
  grid.rect(gp = gpar(col = "black", fill = "grey95"))
  grid.rect(x = unit(q[3], "npc"), width = unit(q[4] - q[3], "npc"),
            just = "left", gp = gpar(fill = "red"))
  p_lwr <- paste(100 * (1 - prob) / 2, "%", sep = "")
  p_mid <- paste(100 * prob, "%", sep = "")
  p_upr <- paste(100 * (1 - (1 - prob) / 2), "%", sep = "")
  grid.text(p_lwr, x = unit(at[1], "npc"), y = unit(.5, "npc"),
            gp = gpar(fontsize = 7))
  grid.text(p_mid, x = unit(at[2], "npc"), y = unit(.5, "npc"),
            gp = gpar(fontsize = 7))
  grid.text(p_upr, x = unit(at[3], "npc"), y = unit(.5, "npc"),
            gp = gpar(fontsize = 7))
  grid.points(pch = 25, size = unit(.5, "lines"), gp = gpar(fill = "black"),
              x = unit(q[3], "npc"), y = unit(1, "npc"))
  grid.points(pch = 25, size = unit(.5, "lines"), gp = gpar(fill = "black"),
              x = unit(q[4], "npc"), y = unit(1, "npc"))
  grid.points(pch = 25, size = unit(.5, "lines"), gp = gpar(fill = "red"),
              x = unit(ifelse(central == "mean", q[1], q[2]), "npc"),
              y = unit(1, "npc"))
  grid.text(round(ifelse(central == "mean", s[1], s[2]), 0),
            gp = gpar(fontsize = 7),
            x = unit(ifelse(central == "mean", q[1], q[2]), "npc"),
            y = unit(1.5, "npc"))
  grid.text(round(s[3], 0), gp = gpar(fontsize = 7),
            x = unit(q[3], "npc"), y = unit(1.5, "npc"))
  grid.text(round(s[4], 0), gp = gpar(fontsize = 7),
            x = unit(q[4], "npc"), y = unit(1.5, "npc"))

  ## histogram
  upViewport(1)
  pushViewport(viewport(y = unit(1, "npc") - unit(4.5, "lines"),
                        x = unit(0, "npc") + unit(4, "lines"),
                        width = unit(.95, "npc") - unit(4, "lines"),
                        height = unit(1, "npc") - unit(8.5, "lines"),
                        just = c("left", "top"), name = "hist"))
  grid.rect(gp = gpar(col = "black"))
  grid.xaxis(at = x_at, label = x_lb,
             gp = gpar(fontsize = 8))
  grid.yaxis(at = .95 * y_at, label = y_lb,
             gp = gpar(fontsize = 8))
  grid.text("Frequency", rot = 90,
            y = unit(0.5, "npc"), x = unit(-3, "lines"))
  grid.text(x$name, y = unit(1, "npc") + unit(2.5, "lines"),
            gp = gpar(fontsize = 15))
  grid.text(xval, y = unit(0, "npc") - unit(3, "lines"))
  grid.polygon(x = rep(x_val, each = 3)[-1],
               y = c(0, y_val, 0),
               gp = gpar(fill = fill, ...))
  grid.lines(x = unit(q[3], "npc"), y = unit(c(0, 1), "npc"))
  grid.lines(x = unit(q[4], "npc"), y = unit(c(0, 1), "npc"))
}